AMD ASU Support

Akshay-Belsare opened the pull request:
Add support for the AMD Application Security Unit (ASU), the on-chip
    Hardware Security Module (HSM) for Versal Gen 2.
    The ASU manages all device-level security services for user
    applications, extending beyond accelerator-centric tasks.
    Its firmware also exposes several software-based cryptographic
    primitives, including:
    - Key transfer
    - RSA authentication (multiple padding schemes)
    - HMAC
    - Key Derivation Function (KDF)
    - Key wrap / unwrap

<!--
    If you are new to submitting pull requests to OP-TEE, then please have a
    look at the list below and tick them off before submitting the pull request.

    1. Read our contribution guidelines:
         https://optee.readthedocs.io/en/latest/general/contribute.html

    2. Read the contribution section in Notice.md and pay extra attention to the
       "Developer Certificate of Origin" in the contribution guidelines.

    3. You should run checkpatch preferably before submitting the pull request.

    4. When everything has been reviewed, you will need to squash, rebase and
       add tags like `Reviewed-by`, `Acked-by`, `Tested-by` etc. More details
       about this can also be found on the link provided above.

    NOTE: This comment will not be shown in the pull request, so no harm keeping
    it, but feel free to remove it if you like.
-->

Akshay-Belsare committed:
drivers: amd: Add ASU support

Add support for the AMD Application Security Unit (ASU), the on-chip
Hardware Security Module (HSM) for Versal Gen 2.
The ASU manages all device-level security services for user
applications, extending beyond accelerator-centric tasks.
Its firmware also exposes several software-based cryptographic
primitives, including:
- Key transfer
- RSA authentication (multiple padding schemes)
- HMAC
- Key Derivation Function (KDF)
- Key wrap / unwrap

Co-developed-by: Harsh Jain <h.jain@amd.com>
Signed-off-by: Harsh Jain <h.jain@amd.com>
Signed-off-by: Akshay Belsare <akshay.belsare@amd.com>
Acked-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswi-linaro commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,422 @@
+	processed = 0;
+	remaining = len;
+	while (remaining) {
+		op.datasize = remaining > ASU_DATA_CHUNK_LEN ?
```
MAX(remaining, ASU_DATA_CHUNK_LEN)?

jenswi-linaro commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,422 @@
+				     ASU_DATA_CHUNK_LEN : remaining;
+		op.opflags = ASU_SHA_UPDATE |
+					   asu_hashctx->shastart;
+		op.dataaddr = (uint64_t)virt_to_phys(data + processed);
```
Is cast needed?

jenswi-linaro commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,422 @@
+		op.datasize = remaining > ASU_DATA_CHUNK_LEN ?
+				     ASU_DATA_CHUNK_LEN : remaining;
+		op.opflags = ASU_SHA_UPDATE |
+					   asu_hashctx->shastart;
```
Please fold up on the line above.

harshjain32 replied:
sure

jenswi-linaro commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,422 @@
+		return TEE_SUCCESS;
+	}
+
+	if (!ctx || (!data && len)) {
```
Can `ctx` ever be NULL?
If we reach this point, we know that len is non-zero, so no need to check it again.

harshjain32 replied:
This is entry point for Driver. not NULL is preventive check of input data.
len == 0 is valid case and driver returning SUCCESS, but
(len && !data) ==> means length is non zero but buffer pointer *data* is NULL, which is BAD parameter.

harshjain32 replied:
Sure, will remove ctx !NULL  check.

jenswi-linaro commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,422 @@
+	struct asu_hash_ctx *asu_hashctx = NULL;
+
+	if (!len) {
+		IMSG("This is 0 len task, skip");
```
`DMSG`?

harshjain32 replied:
Sure

jenswi-linaro commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,422 @@
+	/* Inputs of SHA request */
+	op.dataaddr = 0;
+	op.datasize = 0;
+	op.hashaddr = (uint64_t)virt_to_phys((void *)digest);
```
Is any of the casts needed?

harshjain32 replied:
Can be removed

jenswi-linaro commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,422 @@
+				 uint8_t *digest, size_t len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	struct asu_sha_op_cmd op;
```
Please initialize the locals.

harshjain32 replied:
Sure

jenswi-linaro commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,422 @@
+static TEE_Result asu_hash_do_final(struct crypto_hash_ctx *ctx,
+				    uint8_t *digest, size_t len)
+{
+	struct asu_hash_ctx *asu_hashctx;
```
Please initialize.

jenswi-linaro commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,422 @@
+{
+	struct asu_hash_ctx *asu_hashctx;
+
+	if (!ctx)
```
Can `ctx` be NULL?

harshjain32 replied:
This is entry point for Driver. not NULL is preventive check of input data.

harshjain32 replied:
Sure, will remove ctx !NULL check.

Akshay-Belsare commented on core/drivers/amd/asu/include/asu_client.h:
```diff
@@ -0,0 +1,65 @@
+#define ASU_MODULE_SHA2_ID		1U
+#define ASU_MODULE_SHA3_ID		2U
+
+struct asu_client_params {
```
ASU Client is the code interacting with the ASU Firmware.

jenswi-linaro replied:
So, ASU Client is the driver? Doesn't that make the client part redundant?

Akshay-Belsare replied:
@jenswi-linaro 
The ASU Client is shared code that interfaces with the firmware and exposes a general API set for crypto drivers to use.

Akshay-Belsare commented on core/drivers/amd/asu/asu_main.c:
```diff
@@ -0,0 +1,535 @@
+		ids.id_array[idx] = ASU_UNIQUE_ID_MAX;
+}
+
+/**
```
sure!

jenswi-linaro commented on core/drivers/amd/asu/asu_main.c:
```diff
@@ -0,0 +1,535 @@
+#define ASU_CLIENT_READY		0xFFFFFFFFU
+#define ASU_TARGET_IPI_INT_MASK		1U
+
+#define ASU_GLOBAL_BASEADDR		0xEBF80000U
```
I don't understand. Aren't all physical addresses global in a sense?

ainwood87 commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,422 @@
+		return;
+	asu_hashctx  = to_hash_ctx(ctx);
+	mutex_lock(&asu_shadev->engine_lock);
+	if (asu_hashctx->module == ASU_MODULE_SHA2_ID &&
```
Are sha2_available and sha3_available mutually exclusive?
Can't we just set them like (if using boolean)
```
asu_shadev->sha2_available = (asu_hashctx->module == ASU_MODULE_SHA2_ID );
asu_shadev->sha3_available = (asu_hashctx->module == ASU_MODULE_SHA3_ID );
```

harshjain32 replied:
> Are sha2_available and sha3_available mutually exclusive?
No, H/W has separate instance of sha2 and sha3 engine. Both type of requests can be processed in parallel.
> Can't we just set them like
No, Being shared variable, It will update both sha2_available and sha3_available.

harshjain32 commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,422 @@
+	uint8_t islast;
+	uint8_t opflags;
+	uint8_t shakereserved;
+} asu_ShaOpCmd;
```
It unused. Can be removed.

harshjain32 commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,422 @@
+	TEE_Result ret = TEE_SUCCESS;
+	struct asu_sha_op_cmd op = {};
+	struct asu_client_params *cparam;
+	uint32_t processed;
```
Yes, it can be done. Thanks, will do

jenswi-linaro commented on core/drivers/amd/asu/include/asu_doorbell.h:
```diff
@@ -0,0 +1,66 @@
+ *
+ */
+
+#ifndef _ASU_DOORBELL_H_
```
Can we use `__` instead of `_` as a prefix?

jenswi-linaro commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,462 @@
+		 !asu_shadev->sha3_available)
+		asu_shadev->sha3_available = 1;
+	mutex_unlock(&asu_shadev->engine_lock);
+	asu_hashctx = to_hash_ctx(ctx);
```
This is already done above.

harshjain32 replied:
Will remove

harshjain32 commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,462 @@
+{
+	struct asu_hash_ctx *asu_hashctx = NULL;
+
+	if (!ctx) {
```
Sure, Will update

harshjain32 commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,462 @@
+		asu_shadev->sha3_available = 0;
+	} else {
+		mutex_unlock(&asu_shadev->engine_lock);
+		return TEE_ERROR_NOT_IMPLEMENTED;
```
It will fallback to software.

harshjain32 commented on core/drivers/amd/asu/asu_main.c:
```diff
@@ -0,0 +1,535 @@
+	void *doorbell;
+};
+
+struct asu_ref_to_callback {
```
I missed it, Will remove in next version

harshjain32 commented on core/drivers/amd/asu/asu_main.c:
```diff
@@ -0,0 +1,535 @@
+
+	do {
+		if (io_read32((vaddr_t)asu->global_ctrl) &
+			      ASU_GLOBAL_GLOBAL_CNTRL_FW_IS_PRESENT_MASK) {
```
Sure

harshjain32 commented on core/drivers/amd/asu/asu_main.c:
```diff
@@ -0,0 +1,530 @@
+		return TEE_ERROR_COMMUNICATION;
+	}
+	while (io_read8((vaddr_t)(&bufptr->respbufstatus)) !=
+	      ASU_RESPONSE_IS_PRESENT) {
```
Sure

harshjain32 commented on core/drivers/amd/asu/asu_main.c:
```diff
@@ -0,0 +1,530 @@
+		EMSG("Failed to communicate to ASU");
+		return TEE_ERROR_COMMUNICATION;
+	}
+	while (io_read8((vaddr_t)(&bufptr->respbufstatus)) !=
```
Sure

Akshay-Belsare commented:
@jenswi-linaro Please review at your earliest convenience.

jenswi-linaro commented on core/include/drivers/amd/asu_client.h:
```diff
@@ -0,0 +1,65 @@
+uint8_t alloc_unique_id(void);
+void free_unique_id(uint8_t uniqueid);
+void *asu_update_n_get_ctx(uint8_t unique_id);
+TEE_Result asu_verify_n_get_unique_id_ctx(const void *context,
```
This function is still neither implemented nor called.

jenswi-linaro commented on core/drivers/amd/asu/asu_main.c:
```diff
@@ -0,0 +1,522 @@
+					      int *status)
+{
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	uint8_t freeindex = 0U;
```
`0` should be enough.

jenswi-linaro commented on core/drivers/amd/asu/asu_main.c:
```diff
@@ -0,0 +1,522 @@
+	TEE_Result ret = TEE_ERROR_GENERIC;
+	uint32_t channel_id = 0U;
+	void *asu_shmem = NULL;
+	uint64_t membase = UINT64_C(0);
```
Initializing integers of any size, signed or unsigned, with `0` should be enough. Please fix all occurrences in this PR.

jenswi-linaro commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,444 @@
+	asu_hashctx->uniqueid = ASU_UNIQUE_ID_MAX;
+	mutex_lock(&asu_shadev->engine_lock);
+	if (asu_hashctx->module == ASU_MODULE_SHA2_ID &&
+	    !asu_shadev->sha2_available)
```
If `asu_hashctx->module == ASU_MODULE_SHA2_ID` then `asu_shadev->sha2_available` must be false or there's an inconsistency. There's no need to check `asu_shadev->sha2_available`. If we must, then an `assert()` before setting it to true makes more sense.

jenswi-linaro commented on core/drivers/amd/asu/asu_main.c:
```diff
@@ -0,0 +1,522 @@
+		wfe();
+	}
+
+	*status = bufptr->resp.arg[ASU_RESPONSE_STATUS_INDEX];
```
bufptr->resp.arg[] is an array of `uint32_t`, why not make `status` a `uint32_t` pointer?

jenswi-linaro commented on core/drivers/amd/asu/asu_main.c:
```diff
@@ -0,0 +1,522 @@
+ *
+ * Return: TEE_SUCCESS on successful doorbell trigger
+ */
+static TEE_Result send_doorbell(void)
```
Do you plan for a more complex version of this function that eventually can fail?

harshjain32 commented on core/drivers/crypto/asu_driver/asu_hash.c:
```diff
@@ -0,0 +1,445 @@
+	op.hashaddr = virt_to_phys((void *)digest);
+	op.hashbufsize = len;
+	if (asu_hashctx->shamode == ASU_SHA_MODE_SHA256)
+		op.hashbufsize = ASU_SHA_256_HASH_LEN;
```
Hi,

Actually FW don't not copy hash to digest buffer, but it expects "op.hashaddr" as non zero. It populates hash in response buffer and driver copy it in asu_hash_cb() based on len only.

It will be safe if len is less that hash size.

harshj-xlnx committed:
crypto: asu: Add crypto hash driver

Add support for following Hash algorithms
SHA-256, SHA-384, SHA-512, SHA3-256, SHA3-384, SHA3-512

Signed-off-by: Harsh Jain <h.jain@amd.com>
Signed-off-by: Akshay Belsare <akshay.belsare@amd.com>
Acked-by: Jens Wiklander <jens.wiklander@linaro.org>

jenswi-linaro commented:
Please check for questions or comments I've made that haven't been answered. There are at least two above.
